corehq/messaging/smsbackends/whatsapp/models.py

```diff
@@ -0,0 +1,61 @@
+from __future__ import absolute_import
+from collections import namedtuple
+
+from yowsup.demos.sendclient import YowsupSendStack
+from yowsup.layers import YowLayerEvent
+from yowsup.layers.auth import AuthError
+from yowsup.layers.network import YowNetworkLayer
+
+from corehq.apps.sms.models import SQLSMSBackend
+from corehq.apps.sms.util import clean_phone_number
+from corehq.messaging.smsbackends.whatsapp.forms import WhatsAppBackendForm
+from dimagi.utils import logging
+
+
+YowsupMessage = namedtuple('YowsupMessage', 'phone body')
+
+
+class SQLWhatsAppBackend(SQLSMSBackend):
+
+    class Meta:
+        app_label = 'sms'
+        proxy = True
+
+    @classmethod
+    def get_available_extra_fields(cls):
+        return [
+            'phone_number',
+            'password',
+        ]
+
+    @classmethod
+    def get_api_id(cls):
+        return 'WHATSAPP'
+
+    @classmethod
+    def get_generic_name(cls):
+        return "WhatsApp"
+
+    @classmethod
+    def get_form_class(cls):
+        return WhatsAppBackendForm
+
+    def get_credentials(self):
+        config = self.config
+        return (config.phone_number, config.password)
+
+    def send(self, msg, *args, **kwargs):
+        phone_number = clean_phone_number(msg.phone_number)
+        messages = [YowsupMessage(phone_number, msg.text)]  # msg.text is expected to be Unicode
+
+        stack = YowsupSendStack(self.get_credentials(), messages, encryptionEnabled=True)
+        # stack.start()  # Reimplemented below, but notify on auth failure instead of printing it
+        stack.stack.broadcastEvent(YowLayerEvent(YowNetworkLayer.EVENT_STATE_CONNECT))
+        try:
+            stack.stack.loop()  # TODO: discrete=5, count=X ?
```

Comment:
is there a way to limit how long this loop will run for? we currently try to limit all SMS requests to [5 seconds](https://github.com/dimagi/commcare-hq/blob/7160fe0154f99303bec91720116af06d48341899/settings.py#L1563) otherwise a large backlog of SMS with an unresponsive gateway can back up the queue

Reply:
I think using `discrete` and `count` params might help a lot with that. I was planning to play with those to see what worked well. 

The architecture of the Yowsup library actually fires up a stack of event handlers, connects to WhatsApp, and then allows you to send a whole lot of messages. The way I've implemented it here, it recreates the stack and reconnects to WhatsApp for each message. That keeps the code very simple, but it's not ideal. If we want to send more than one message, we should reuse the same connection. ... If this PR is useful, it might be worth considering.